webrtc/stun/stun_connection.nim

```diff
@@ -0,0 +1,249 @@
+  var unknownAttr: seq[uint16]
+  for attr in msg.attributes:
+    let typ = attr.attributeType
+    if typ.isRequired() and typ notin StunAttributeEnum:
+      unknownAttr.add(typ)
+  if unknownAttr.len() > 0:
+    res.attributes.add(ErrorCode.encode(ECUnknownAttribute))
+    res.attributes.add(UnknownAttributes.encode(unknownAttr))
+    return some(res)
+
+  return none(StunMessage)
+
+proc isFingerprintValid*(msg: StunMessage): bool =
+  # Returns true if Fingerprint is missing or if it's valid.
+  # Returns false otherwise.
+  let fingerprint = msg.getAttribute(Fingerprint)
+  if fingerprint.isNone():
+    return true
+  if msg.attributes[^1].attributeType != AttrFingerprint:
+    # Fingerprint should always be the last attribute.
+    return false
+  let
+    copyWithoutFingerprint = StunMessage(
+      msgType: msg.msgType,
+      transactionId: msg.transactionId,
+      attributes: msg.attributes[0 ..< ^1]
+    )
+    encodedCopy = copyWithoutFingerprint.encode(@[])
+  return fingerprint == StunMessage.decode(encodedCopy).getAttribute(Fingerprint)
+
+# - Stun Messages Handler -
+
+proc stunMessageHandler(self: StunConn) {.async: (raises: [CancelledError]).} =
+  # Read indefinitely Stun messages from stunMsgs queue.
+  # Sends a BindingResponse or BindingResponseError after receiving a BindingRequest.
+  while true:
+    let message = await self.stunMsgs.popFirst()
+    try:
+      let decoded = StunMessage.decode(await self.stunMsgs.popFirst())
+      if not decoded.isFingerprintValid():
+        # Fingerprint is invalid, the StunMessage received might be a false positive.
+        # Move this message to the `dataRecv` queue
+        await self.dataRecv.addLast(message)
+        continue
+      if decoded.msgType == StunBindingErrorResponse:
+        trace "Received a STUN error", decoded, remote = self.raddr
+        continue
+      elif decoded.msgType == StunBindingResponse:
+        # TODO: Process StunBindingResponse doesn't seem necessary for libp2p-webrtc-direct.
+        # Some browsers could be uncooperative. In that case, it should be implemented.
+        # It should be implemented for libp2p-webrtc.
+        continue
+      elif decoded.msgType == StunBindingRequest:
+        let errorOpt = self.checkForError(decoded)
+        if errorOpt.isSome():
+          let error = errorOpt.get()
+          await self.conn.write(self.raddr, error.encode(@[]))
+          continue
+
+        let
+          bindingResponse = self.getBindingResponse(decoded)
+          usernameAttr = decoded.getAttribute(UsernameAttribute).get()
+          password = self.passwordProvider(usernameAttr.username)
+        await self.conn.write(
+          self.raddr,
+          bindingResponse.encode(password)
+        )
+    except SerializationError as exc:
+      warn "Failed to decode the Stun message", error=exc.msg, message
+    except WebRtcError as exc:
+      trace "Failed to write the Stun response", error=exc.msg
+
+proc init*(
```

Comment:
It should be `new` as StunConn is a `ref`.

Reply:
done